### DISCUSSION TITLE
graphql_flutter Query widget throws error when loading the widget

### DISCUSSION THREAD

**Discussion by sumchans** [zino-app/graphql-flutter#1326] [Q&A] [answered]
I am not sure what I am doing wrong here. I get an error when I query my graphql endpoint using the Query widget of the graphql_flutter package.

Here is my query -
```
const String findContacts = """ {
  query (\$contacts:[String]){
  findContacts(contacts:\$contacts) {
      status
      error
      users {
          id
          userPhoneNumber
          userDisplayName
      }
    }
  }
}  
""";
```
Here is my implementation of the Query widget-

```
    Query(
      options: QueryOptions(
        document: gql(graphql_string.findContacts),
        variables: const <String, dynamic>{
          'contacts': ["2612", "1212"],
        },
      ),
      builder: (QueryResult result,
          {VoidCallback? refetch, FetchMore? fetchMore}) {
        //if (result.data != null) {
        return ElevatedButton(
          onPressed: () {
            print(result.data!["findContacts"]["users"]);
          },
          child: const Text('FETCH CONTACTS'),
        );
      },
    ),
```

Below is the error I see-

![image](https://github.com/zino-hofmann/graphql-flutter/assets/33538243/a87e22e5-b803-4544-9555-08fce8697a7a)

**Comment by vincenzopalazzo** [comment]
Try the following as mentioned in the docs https://github.com/zino-hofmann/graphql-flutter/tree/main/packages/graphql

```dart
const String findContacts = r''' {
  query ($contacts:[String]){
  findContacts(contacts:\$contacts) {
      status
      error
      users {
          id
          userPhoneNumber
          userDisplayName
      }
    }
  }
}  
''';
```

**Reply by sumchans** [reply]
Hello @vincenzopalazzo.

I tried the above code, just renamed the variable and the schema.
```
const String findContactsOnBol = r''' {
  query ($contacts:[String]){
  findContactsOnBol(contacts:\$contacts) {
      status
      error
      users {
          id
          userPhoneNumber
          userDisplayName
      }
    }
  }
}  
''';
```

But I still get an error as below-
<img width="551" alt="image" src="https://github.com/zino-hofmann/graphql-flutter/assets/33538243/db18a92c-b4f7-46b5-ba75-efc44ec49635">

**Reply by vincenzopalazzo** [reply]
you need to remove all the `\$` with `$`

**Reply by sumchans** [reply]
I tried that as well, forgot to mention -
```
const String findContactsOnBol = r''' {
  query ($contacts:[String]){
  findContactsOnBol(contacts:$contacts) {
      status
      error
      users {
          id
          userPhoneNumber
          userDisplayName
      }
    }
  }
}  
''';
```

When I do that it points me to the below line of code on the parser.dart file-
<img width="479" alt="image" src="https://github.com/zino-hofmann/graphql-flutter/assets/33538243/37c6387c-b180-4265-b39f-f10c164ef5d5">

**Comment by vincenzopalazzo** [comment]
This is an exception; you should print the exception.

I do not use vscode no I do not know what it is mean, but this is an exception

**Reply by sumchans** [reply]
This is what the errorMessage says-
<img width="794" alt="image" src="https://github.com/zino-hofmann/graphql-flutter/assets/33538243/f1859f84-2400-48ad-a7c4-4f3fd0bdaf3a">

**Reply by sumchans** [reply]
@vincenzopalazzo any idea on what's going on with the code.

**Reply by vincenzopalazzo** [reply]
there is no variable name in your example, please take time to reflect on what you want to do, maybe you are not passing the variable in the query call?

**Reply by sumchans** [reply]
The query works as expected if I put it on to the graphql playground. I do have added the variables property just below the document property of the QueryOptions object. For testing purposes, I have added hardcoded values for the variable. The query takes in an array of strings and returns matching documents. Below is the schema for the above query-
```
type Query {
        hello: String!
        users: Result!
        user(userPhoneNumber:ID!): Result!
        findContactsOnBol(contacts: [String]): Result!
    }

    type Mutation {
        userCreate(user: UserInput!): Result!
        userUpdate(id: ID! user: UserInput): Result!      
    }

    type User {
        id: ID!
        partitionKey: String 
        userPhoneNumber: ID      
        userDisplayName: String
        avatarUrl: String
        createdAt: String 
        updatedAt: String
        userStatus: String
        userCustomMessage: String
    } 

    input UserInput {
        id: ID
        partitionKey: String
        userPhoneNumber: ID
        userDisplayName: String
        avatarUrl: String
        createdAt: String
        updatedAt: String
        userStatus: String
        userCustomMessage: String
    }

    type Result {
        status: Int
        error: String
        user: User
        users: [User]
    }
```

**Reply by vincenzopalazzo** [reply]
just working in the playground does not mean that this should work in the dart code; if the playground adds a variable or modification to the code, you do not know.

Try to use the same query with curl and see what it is the result

**Reply by sumchans** [reply]
I just dont have any idea on what is wrong here. I have tried my best, I just cant seem to find what's wrong here.

**Reply by vincenzopalazzo** [reply]
Try to run the same command with curl from the command line and see what happens, I am pretty sure that there is something that it is going on

**Reply by sumchans** [reply]
It works on curl- here is the snip -
<img width="1046" alt="image" src="https://github.com/zino-hofmann/graphql-flutter/assets/33538243/42be0b68-a16e-435f-8064-8a7b2093b0af">

**Reply by vincenzopalazzo** [reply]
No sorry this is not the same query, yours is a `query { .... }` but in flutter you are using another kind of query that is `query Foo(args...) { ... }`

Please take a look at the spec https://graphql.org/learn/queries/#operation-name IMHO your query this not correct and this should solve the problem

```dart
const String findContactsOnBol = r''' {
  query FindContactsOnBol($contacts: [String]) {
    findContactsOnBol(contacts: $contacts) {
        users {
            userPhoneNumber
        }
      }
    }
}  
''';
```

**Reply by sumchans** [reply]
@vincenzopalazzo the query above also gives me the same error. I noticed that if I provide a query to the Query widget that does not take any parameter works, but any query that takes in parameters errors out. 

The below query does not take any parameter-
```
const String getUsers = """
query {
  users {
    users {
      userPhoneNumber
    }   
  }
}
""";
```

Here is the Query widget -

```
            Query(
              options: QueryOptions(
                document: gql(graphql_string.getUsers),
                onError: (error) => print(error),
                onComplete: (data) => print(data),
                // variables: <String, dynamic>{
                //   'contacts': contacts,
                // },
              ),
              builder: (QueryResult result,
                  {VoidCallback? refetch, FetchMore? fetchMore}) {
                //if (result.data != null) {
                return ElevatedButton(
                  onPressed: () {
                    //print(result.data!["user"]["user"]["userPhoneNumber"]);
                  },
                  child: const Text('FETCH CONTACTS'),
                );
              },
            ),
```

**Reply by vincenzopalazzo** [reply]
if without parameters works there is some error with your code, but I lost the context so if you think that this is a bug, please make a reproducer repository that I can test on my machine.

But if I was in you I was going to read the docs of the graphql_flutter to check what your code is doing wrong i am pretty sure that the bug is in your code.

**Reply by sumchans** [reply, marked as answer]
I was finally able to fix it. This is way how I got it worked -

```
const String findContactsOnBol = """
  query(\$contacts:[String]) {
    findContactsOnBol(contacts:\$contacts) {
      status
      error
      users {
          id
          userPhoneNumber
          userDisplayName
      }
    }
  }
""";
```

**Reply by vincenzopalazzo** [reply]
you should be able to remove \ if you use a raw string. Happy that you find a solution

**Comment by ndelanou** [comment]
Have you tried naming your query ?

```
const String findContacts = r""" {
  query yourQueryName($contacts:[String]) {
    findContacts(contacts:$contacts) {
      status
      error
      users {
        id
        userPhoneNumber
        userDisplayName
      }
    }
  }
}  
""";
```

**Reply by sumchans** [reply]
@ndelanou yes I have already tried that.
